default-config.json

```diff
@@ -4,7 +4,7 @@
     "stop": "Cadence stop",
     "help": "Cadence help",
     "nowplaying": "Cadence now playing",
-    "search": "Cadence search ",
+    "search": "Cadence search <query>",
```

Comment:
If we're adding parameter syntax (and hoping the general public understands to `Cadence search railgun`, not `Cadence search <railgun>`), should we also do so for request? That might be too intricate - With OSR included, request may operate on either a query or a number from a search, and suddenly the help text to explain that is long and probably not all that helpful. Hm.

Thoughts?

Reply:
Might it be clearer if we code-block the "\<query\>" part? Again, to people not used to parsing the output of `<command> -h`:

1. Cadence search \<query\>: Search the Cadence database with a \<query\>
2. Cadence search `<query>`: Search the Cadence database for `<query>`

What do you think?

Reply:
First, I think the parameter syntax using the `< >` symbols are fine. Other discord bots describe their argument usage the same way. If we want to, maybe we can say "Search the Cadence database with a \<query> (eg. `Cadence search railgun`)"

Second, I did think about using `Cadence request <selection>`, but as you mentioned about OSR, I wasn't sure if the word "selection" accurately described **both** a selection number (for two-step) or a song/title/artist (one-step). Like before, maybe the simplest solution is to add an example? "Request a song \<selection> (eg. `Cadence request 1` or `Cadence request railgun`)"

Reply:
Oddly enough, for the lead dev of a Discord bot, I don't spend much time looking at Discord bots that aren't my own. If this is standard, I withdraw my commentary on it.

Yeah, and the more I think about it the more complicated describing the syntax is - OSR means that request takes a selection number OR a search query.

And even though search queries are superficially simple, if we really wanted to document things throughly, we would be describing the special request forms Cadence supports serverside - They're not really CadenceBot's problem (okay, they're literally not CadenceBot's problem, since CadenceBot doesn't know or care about the details of how a search happens), but they are a significant way that searches can be done. But that means that the full spec for the "selection" parameter to request (without pointing at an external command) looks like:

> \<selection\> can be either a number or a string.
>
>If a number, request the associated song from the most recent results of `search`, `library`, or a `request` with a string selection that could not be narrowed to one song (see below).
>
>If a string, then will perform a search for \<selection\>. Search can have a number of special forms:
> 1. If \<selection\> matches "songs named \<term\>", then search for songs whose title matches \<term\>
> 2. If \<selection\> matches "songs by \<term\>", then search for songs by an artist that matches \<term\>
> 3. If \<selection\> matches "\<term\> songs", then search for songs whose genre matches \<term\>
> 4. If \<selection\> matches "songs from \<term\>", then search for songs which either came out in the year described by \<term\> or which are from an album which matches \<term\>
> 5. If \<selection\> matches "songs released in \<year\>", then search for songs which came out in the year \<year\>
> 6. If \<selection\> matches "songs in \<term\>", then search for songs from an album which matches \<term\>  
>
> If \<selection\> does not match one of these forms, then search for songs whose title matches \<selection\>, or which are by an artist that matches \<selection\> 
> After search, attempt to narrow down the results as far as possible, by attempting to discard non-exact matches or identify common patterns in search terms. If it is possible to select one result which is obviously better than any other (for example, if the search term only matches one result to begin with), then submit a request for that song. Otherwise, reply with the smallest list of results that could not be selected from, and ask for a second request to be submitted to request one of them.

.... And while this is good information to have, and we don't really document it in a user-available way anywhere, the shorttext for `Cadence help` is hardly the appropriate place. I think the original idea behind how I wrote the helptext for search and request is 'it should be sort-of intuitive what you give to these commands (Especially with OSR), so CadenceBot should figure out how to handle whatever comes out' - IE that it's a sort of light NLP to say `Cadence request only my railgun`, because that's sort of intuitively how one constructs that command. Although it would make sense to then support `Cadence search for only my railgun` (discarding the 'for') - But then it's difficult to distinguish when we should leave the term (`Cadence search for crying out loud`) from when we should discard it.

So I'm not opposed to examples per se, it's just that depending on how in detail we describe things the full mechanism of these commands is difficult. Not to mention `play` now having autoselect logic backing it..

Reply:
Perhaps what we should do is simply leave the text as `Cadence search <term>` and `Cadence request <target>` (or \<selection\>), and throw in a `Cadence help <command>` syntax - ie `Cadence help` gives these short messages, and then `Cadence help request` gives the description of how exactly one can request, and how overloaded request is (because seriously, the source for `Cadence request` is huge). Although it would probably say "See `Cadence help search` for more information" instead of fully describing the search forms as I did above...

Once upon a time, I would have protested at least quietly how much this would bloat the default-config. Now that there are custom commands, I don't care anymore! So if you think this is a good idea, let's leave it out of this PR and file a new issue to support in-depth helptext.

Reply:
In-depth helptext - Just a thought - Could also then include a `Cadence help other` or something, that prints an autogenerated helptext of all the loaded custom commands. Obviously we're not going to print out the full spec of the responses, but perhaps something like:

> - Cadence introduce yourself: Prints a static response
> - Cadence flip a coin: Prints a randomly selected response.
> ...
> - Cadence resolve the debate between \<mention 1\> and \<mention 2\>: Prints a randomly selected response mentioning the 2 users mentioned by the command

That might take a little fun out of discovering what silly things I write in (And would also be a little annoying with all the variations of "Thanks, Cadence!" that are currently loaded), but otherwise there's no real way to find them.

I suppose then the question becomes whether these commands are more like an 'easter egg' (Which I think was the original design, where one would happen upon a command by accident more often than one triggers it on purpose), or more like a real feature (Which is basically how we use most of them "in the field"). Hmm - Food for thought I suppose.